torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+    return dataset_;
+  }
+
+  void read_from_file() {
+#ifdef CTF_DEBUG
+    size_t read_count = 0;
+#endif
+
+    do {
+#ifdef CTF_DEBUG
+      std::cout << "Read count: " << ++read_count << " starting at "
+                << reader_->get_position() << std::endl;
+#endif
+      // CTF files start with valid alpha-numeric characters
+      if (is_non_printable(reader_->peek_char())) {
+        std::string error_msg(
+            "Non printable character anon print CTF file at position " +
+            std::to_string(reader_->get_position()) + "(" +
+            std::to_string(static_cast<int>(reader_->peek_char())) + ")");
+#ifdef CTF_DEBUG
+        std::cout << error_msg << std::endl;
+#endif
+        throw std::runtime_error(error_msg);
+      }
+
+      // There can be an explicit sequence ID at the beginning of the line or
+      // the last known is used implicitly
+      CTFSequenceID sequence_id;
+      bool is_new_sequence = get_sequence_id(sequence_id);
+
+      while (!is_eol(reader_->peek_char())) {
+        // After the sequence ID, there can be many input streams/comments
+        if (!get_input_stream(sequence_id, is_new_sequence)) {
+          if (!discard_comment()) {
+            std::string error_msg(
+                "Invalid CTF File. Neither a CTF Value nor a "
+                "CTF Comment was found at position " +
+                std::to_string(reader_->get_position()));
+#ifdef CTF_DEBUG
+            std::cout << error_msg << std::endl;
+#endif
+            throw std::runtime_error(error_msg);
+          }
+        }
+      }
+      // Discard EOL
+      reader_->get_char();
+    } while (reader_->can_read());
+  }
+
+ private:
+  CTFParser() = delete;
+  DISALLOW_COPY_AND_ASSIGN(CTFParser);
+
+  bool get_sequence_id(CTFSequenceID& sequence_id) {
+#ifdef CTF_DEBUG
+    // For logging purposes
+    size_t initial_pos = reader_->get_position();
+#endif
+
+    // Flag to identify when a new Sequence ID is found
+    bool is_new = false;
+
+    // idx will be used to iterate through scratch_ for local string parsing
+    size_t idx = 0;
+
+    // Sequence ID must start with a digit
+    char c = reader_->peek_char();
+    if (!is_digit(c)) {
+#ifdef CTF_DEBUG
+      std::cout << "Not a Sequence ID at position " << initial_pos << std::endl;
+#endif
+      if (has_initial_sequence_id_) {
+        sequence_id = previous_sequence_id_;
+#ifdef CTF_DEBUG
+        std::cout << "Using previous Sequence ID (" << previous_sequence_id_
+                  << ")" << std::endl;
+#endif
+      } else {
+        is_new = true;
+        sequence_id = previous_sequence_id_ + 1;
+
+#ifdef CTF_DEBUG
+        std::cout << "Incremented previous Sequence ID (" << sequence_id << ")"
+                  << std::endl;
+#endif
+      }
+      previous_sequence_id_ = sequence_id;
+      return is_new;
+    }
+
+    // Get all consecutive digits
+    while (is_digit(reader_->peek_char())) {
+      c = reader_->get_char();
```

Comment:
Instead of reading one char at a time, we can make it more efficient by reading the whole substring till it hit the delimiter. 
There is an method called strtok that takes an input of char*, a list of delimiters and split the whole string. 
http://www.cplusplus.com/reference/cstring/strtok/

Maybe we can provide a readInteger method in the reader class, which calls strtok. Now the refill() may need more attention, but given the frequency of calling get_char(), and calling refill(), we can still get a reasonable perf gain.

Reply:
Wouldnt strtok() read char by char too, to detect the token?

Reply:
How strtok() is implemented is depending on std library, but the advantage of using it here is that, during we read the whole number string, we don't need to constantly updating previous_char_, checking rewinded_char_ etc, which is get_char() does every time.